### DISCUSSION TITLE
Tree component

### DISCUSSION THREAD

**Discussion by avanslaars** [cengage/react-magma#23] [Ideas]
Let's start discussing the needs and design around a tree view component and then we can determine the effort and where it falls in the roadmap.

**Comment by avanslaars** [comment]
I'll start with the assumption that whatever we build should probably support a recursive data structure that allows nesting to an arbitrary depth. We'll probably also want a lot of flexibility in the way individual items are rendered, so default components that can be replaced with custom components like our Select and Combobox components.

**Reply by paulwvaillancourt** [reply]
Yes, we would need to be able to render components with multiple checkboxes per item like:

![Screen Shot 2020-12-11 at 2 39 26 PM](https://user-images.githubusercontent.com/59023821/101947576-b3205e00-3bbe-11eb-83d6-dbf79f7deeb2.png)

Here we could not only select the item, but attributes within the item as well. In this case, select the item, and whether it would be included in a gradebook. This would, of course, be a custom component. But just an example of our need.

**Comment by orion-cengage** [comment]
I've taken some time to continue the design of a Tree View component for Magma and I'm posting that here for the purposes of discussion.

To be clear, this first round is for a general Tree View control which is used to show a hierarchical list of data with some basic functionality.  When creating something like left-navigation with a hierarchy, I imagine that would probably be a modified/extended version of the basic component, but with navigation-specific requirements accounted for.  Same for the example from @paulwvaillancourt above, that looks like a tree view within a table, which has different requirements that need to be accounted for.

**inVision:** https://cengage.invisionapp.com/share/6CZXHAZPUJ5

But for now, we're starting with the basic component with the following capabilities:

1. Technically it should support an infinite number of levels, but great care should be taken in the design and implementation to not make it hard to use or unnecessarily complex
2. Optional ability to make the levels collapse and expand, including whether a level is already open when the component is loaded
3. Optional ability to select items in the hierarchy with checkboxes
4. Optional ability to display icon on each item in the hierarchy

**Comment by orion-cengage** [comment]
I updated the inVision URL above.  It contains new screens with refinements to the basic component, along with the following additions:

- Drag and drop
- Including additional components within the rows that can impact the overall height of the row
- Applying tree view to side navigation
- Applying tree view inside a table
